Added TexpadTeX cases

vallettaventures opened the pull request:
We've been developing TexpadTeX as an integrated typesetter for Texpad for some time now. It was forked from KerTeX originally, but since we have rewritten the original eTeX WEB source in C++, with a new Pascal runtime, and extended it to support a number of more modern packages, such as `fontspec` and `pgf`. We have also patched a number of standard LaTeX packages such that they backend correctly onto TexpadTeX.

We'd like to merge a number of our package changes upstream now, so we don't need to rebase the changes every time the package changes. As a first step we are hoping to add an `iftexpadtex` macro into the new `iftex` package. If that were to be in place we could approach individual package authors with patches to make them compatible with TexpadTeX.

This patch uses the `TexpadTeXMode` command, which will be in the next version of TexpadTeX, released with the iOS and macOS apps, but it won't work with the currently released version. We can send a beta of the current version if required for testing. Also, as TexpadTeX uses an Apple PDF library, we don't have a binary right now that could be used on any platform for command line testing. We are working on a common PDF library, after which our intention is to create a binary download of TexpadTeX with the same interface as other TeXs.

Commit:
Added TexpadTeX cases

This adds the following commands to cover TexpadTeX

- \iftexpadtex, \ifTexpadTeX
- \iftexpadtexunicode, \ifTexpadTeXUnicode
- \iftexpadtexeightbit, \ifTexpadTeXEightBit
- \RequireTexpadTeX, \RequireTexpadTeXUnicode, \RequireTexpadTeXEightBit

Please note that TexpadTeX can run in one of two modes

- 8-bit mode roughly similar to pdfTeX
- Unicode mode roughly similar to XeTeX & LuaTeX

davidcarlisle commented:
Thank you for your interest in iftex.

We are slightly concerned about adding three new ifs plus their mixed case aliases and three new require. We are aware of several TeX implementation projects and if we add nine new documented commands for each, then  the "small" iftex may end up with a lot of code that we can't really test (our CI test environment runs on a public linux travis CI build on github).

Note that iftex was never intended to include all possible tests. For example it has some basic tests for ptex, but the ifptex package is still separately maintained and has several more tests for specific Japanese typesetting modes.

Here I wonder if we could just have `\iftexpadtex` (and mixed case and Require versions of that)
but then overload the existing `\iftutex` to distinguish your Unicode  version.

If you support `\Umathchardef` then `\iftutex` will already be true in your Unicode version (this would be best) if you don't support  Unicode math yet then the iftex package could set `\iftutex`  true if  `\TexpadTeXMode=1`

That would allow portable documents of the form

```
\usepackage{iftex}
\iftutex
\usepackage{fontspec}
...
\else
\usepackage{times}
\fi
```

not needing a texpadtex-specific Unicode test.

So a diff something like

```
diff --git a/iftex.sty b/iftex.sty
index 8306f86..bedb14b 100644
--- a/iftex.sty
+++ b/iftex.sty
@@ -88,6 +88,7 @@
 \IFTEX@protected\def\RequireAlephTeX{\IFTEX@Require\ifalephtex{Aleph}\fi}
 % tutex == has \Umathchardef == XeTeX or Lua(HB)TeX currently
 \IFTEX@protected\def\RequireTUTeX{\IFTEX@Require\iftutex{LuaTeX or XeTeX}\fi}
+\IFTEX@protected\def\RequireTexpadTeX{\IFTEX@Require\iftexpadtex{TexpadTeX}\fi}
 
 
 % As a matter of policy over-write any existing \if*tex macro and set
@@ -201,6 +202,15 @@
 \IFTEX@let{AlephTeX}{alephtex}
 
 
+%texpadtex
+\begingroup\expandafter\expandafter\expandafter\endgroup
+\expandafter\ifx\csname TexpadTeXMode\endcsname\relax
+   \IFTEX@let{texpadtex}{false}
+\else
+   \IFTEX@let{texpadtex}{true}
+\fi
+\IFTEX@let{TexpadTeX}{texpadtex}
+
 % tutex (LuaTeX or XeTeX)
 \begingroup\expandafter\expandafter\expandafter\endgroup
 \expandafter\ifx\csname Umathchardef\endcsname\relax
@@ -228,6 +238,10 @@ end
   \ifnum\OpMode=3 %
     \pdftrue
   \fi
+\else
+  \iftexpadtex
+    \pdftrue
+  \fi
 \fi
 \else
   \ifnum\pdfoutput>0 %
```



Not directly related to the pull request to the iftex package we would be interested to know what other changes to packages you need to ask for?

In an ideal world few packages would need engine testing, the core engine differences should be isolated in the latex2e graphics-def files as used by color and graphics packages, or in the expl3 l3backend code.  Other packages if they then use the abstracted interfaces should be able to use the supplied functions (for color, scaling, graphics, etc) without knowing engine details. We know that in practice it doesn't always work that way but we would be interested to see where you are needing to make changes.


In particular differences in Unicode font handling between xetex and luatex is already a major issue in our testing framework and for users
that have to switch between engines.
We therefore hope that the font handling appears to a package such as
fontspec as "xetex-like or "luatex-like" (most likely the long-term go
forward model)  and *not* as completely new variant.

vallettaventures commented:
Thanks for replying.

Yes, that makes sense. In the case that `\iftexpadtex` returns true, the `\TexpadTexMode` could be used directly so there is no need for the extra commands. We do not support `\Umathchardef` yet, but it is on the list. `\iftutex` could use `\TexpadTeXMode` for now, then drop that test when `\Umathchardef` is supported? If that is ok, I'll update the pull request accordingly.

To answer your more general question.  Our full list of package patches is

- **base** `fonttext.ltx` and `load-unicode-data` currently use `\Umathchar` to test for unicode engines. Would the new `iftutex` command be useable that early? If so, I could submit a pull request for it?
- **graphics** We have a `texpadtex.def` right now. The intent is to clean up and document the TexpadTeX commands before we submit a pull request.
- **l3backend** we haven't needed to override this yet, but it is on the list to do once the TexpadTeX commands are cleaned up.
- **crop** we have a minor patch for setting paper dimensions. The package hasn't changed on CTAN for 17 years, but I think it would be possible to backend `crop` onto the `papersize` package instead. If it were possible to get a patch merged I'd rather do that than something TexpadTeX specific.
- **mathspec** This package is only compatible with XeTex, not LuaTeX.  We have a patch to make it compatible with TexpadTeX.
- These have driver files, so we've appended a TexpadTeX one.  I don't think there is an obvious way to backend any of these onto l3backend code.
    - **pdfpages**
    - **pgf**
    - **hyperref**
- These packages have either a LuaTeX or a XeTeX implementation, and we've added a third.  I understand your point about not wanting to merge that though. It is not hard for us to change the new TeX commands we've exposed, so I can look into making the interface either XeTeX-like, or LuaTeX-like when we get there.
    - **polyglossia**
    - **fontspec**

josephwright commented:
We don't necessarily need to load `iftex`  for format-building stuff, that's actually better done by explicitly coding in the tests. So adjusting `load-unicode-data` or whatnot is easy enough. What is needed, though, is a spec for what is supported (particularly if you look at things like `l3backend`). I'm wondering how you are supporting Unicode without `\Umathchar`, as that's a core part of Unicode math mode.

What is available in terms of documentation? Ideally, macro programmers work from documentation from the engine developers that's reasonably detailed. (Yes, the real world isn't like that, but I can dream). I'm wondering about your coverage of the 'pdfTeX utilities', as these are used for bits of `expl3`, for example: not directly relevant here, but likely to be important in terms of things like `l3backend` (as it's got to actually load!).

davidcarlisle commented:
@vallettaventures  OK thanks.

If you could modify the PR just to define the one \if (and its mixed case alias) and \Require, and to set \iftutex then that would be good.

On the question of whether we should have iftex.sty available in the format or otherwise make something like \iftutex available at format creation time rather than \ifx\Umath.. tests, it isn't the case now but doing something along those lines in future isn't impossible. However that decision need not hold up adding to iftex.sty now.

u-fischer commented:
> * Would the new `iftutex` command be useable that early? If so, I could submit a pull request for it?

In the kernel we are loading expl3 which offers tests like `\sys_if_engine_luatex:` and similar. Currently expl3 is loaded very late and so can't be used in fonttext.ltx and similar, but this can probably be changed.  



> * These have driver files, so we've appended a TexpadTeX one.  I don't think there is an obvious way to backend any of these onto l3backend code.

hyperref (and related packages like bookmark) will be based on the l3backend code in the feature. There is no plan here to add more hyperref specific drivers. Instead there will be one generic driver (and probably a few older one for compability) which will use the l3backend code.

Commit:
Fixes for TexpadTeX checks

- Removed unnecessary tests for TexpadTeX Unicode mode
- Updated `iftutex` to detect TexpadTeX's Unicode mode

vallettaventures commented:
@davidcarlisle I've updated the pull request along those lines, let me know if there is anything further I should alter.

vallettaventures commented:
@josephwright Yes, we definitely need to sort the documentation. Up until now we've deliberately avoided publicly documenting the commands we added as they were not finalised and we didn't want to maintain backward compatibility with them. Now that we are happy with them, we will produce proper documentation before submitting any further pull requests.

We would have done so before this one, but we are writing a new help system on our website, and it is not quite ready to be deployed yet.

josephwright commented:
@vallettaventures  Any chance of a 'pre-print' by email to either me or the team? Obvious that would be confidential,  but it would be good to see the direct before it's public and (presumably) set in stone.

vallettaventures commented:
@josephwright Sure, there is nothing really confidential in there, its just a case of making sure we don't publicly promise to continue supporting macros we plan to remove.  What email should I use? If you'd rather not post that publicly for spam reasons, please email duncan@vallettaventures.com

josephwright commented:
@vallettaventures joseph.wright@morningstar2.co.uk

davidcarlisle commented:
Thanks I merged it locally, I moved the ifTuTeX alias later after the new code had potentially changed \iftutex, and bumped the version number, but the copy in master should work for you now, could you confirm, thanks.

vallettaventures commented:
Yes, that is great, thank you.  We've got some work to do on our graphics system now, then when that is cleaned up, we'll be in touch with a pull request for that repo.
